test/scDAI.t.sol

```diff
@@ -0,0 +1,446 @@
+
+    function test_constructor() public {
+        assertEq(address(vault.asset()), C.SDAI);
+        assertEq(address(vault.scWETH()), address(wethVault), "weth vault");
+        assertEq(address(vault.priceConverter()), address(priceConverter), "price converter");
+        assertEq(address(vault.swapper()), address(swapper), "swapper");
+
+        assertEq(weth.allowance(address(vault), address(vault.scWETH())), type(uint256).max, "scWETH allowance");
+    }
+
+    function test_depositDai(uint256 daiAmount) public {
+        daiAmount = bound(daiAmount, 1e18, 100_000_000e18);
+        deal(C.DAI, address(this), daiAmount);
+        dai.approve(address(vault), daiAmount);
+        vault.depositDai(daiAmount, address(this));
+
+        uint256 expectedShares = sDai.convertToShares(daiAmount);
+
+        assertEq(vault.balanceOf(address(this)), expectedShares, "shares dont match");
+    }
+
+    function test_rebalance() public {
+        uint256 initialBalance = 1_000_000e18;
+        uint256 initialDebt = 100 ether;
+        deal(address(sDai), address(vault), initialBalance);
+
+        bytes[] memory callData = new bytes[](2);
+        callData[0] = abi.encodeWithSelector(scDAI.supply.selector, spark.id(), initialBalance);
+        callData[1] = abi.encodeWithSelector(scDAI.borrow.selector, spark.id(), initialDebt);
+
+        vault.rebalance(callData);
+
+        assertEq(vault.totalDebt(), initialDebt, "total debt");
+        assertEq(vault.totalCollateral(), initialBalance, "total collateral");
+
+        _assertCollateralAndDebt(spark.id(), initialBalance, initialDebt);
+
+        assertEq(wethVault.balanceOf(address(vault)), initialDebt, "scETH shares");
+    }
+
+    function testFuzz_rebalance(uint256 supplyOnSpark, uint256 borrowOnSpark) public {
+        uint256 floatPercentage = 0.01e18;
+        vault.setFloatPercentage(floatPercentage);
+
+        supplyOnSpark = bound(supplyOnSpark, 100e18, 1_000_000e18);
+
+        uint256 initialBalance = supplyOnSpark.divWadDown(1e18 - floatPercentage);
+        uint256 minFloat = supplyOnSpark.mulWadDown(floatPercentage);
+
+        borrowOnSpark = bound(
+            borrowOnSpark,
+            1e10,
+            priceConverter.sDaiToEth(supplyOnSpark).mulWadDown(spark.getMaxLtv() - 0.005e18) // -0.5% to avoid borrowing at max ltv
+        );
+
+        deal(address(sDai), address(vault), initialBalance);
+
+        bytes[] memory callData = new bytes[](2);
+        callData[0] = abi.encodeWithSelector(scDAI.supply.selector, spark.id(), supplyOnSpark);
+        callData[1] = abi.encodeWithSelector(scDAI.borrow.selector, spark.id(), borrowOnSpark);
+
+        vault.rebalance(callData);
+
+        _assertCollateralAndDebt(spark.id(), supplyOnSpark, borrowOnSpark);
+        assertApproxEqAbs(vault.totalAssets(), initialBalance, 1e10, "total assets");
+        assertApproxEqAbs(vault.sDaiBalance(), minFloat, vault.totalAssets().mulWadDown(floatPercentage), "float");
+    }
```

Comment:
```suggestion
    function testFuzz_rebalance(uint256 supplyOnSpark, uint256 borrowOnSpark) public {
        uint256 floatPercentage = 0.01e18;
        // -10 to account for rounding error difference between debt vs invested amounts
        vault.setFloatPercentage(floatPercentage - 10);

        supplyOnSpark = bound(supplyOnSpark, 100e18, 1_000_000e18);
        borrowOnSpark = bound(
            borrowOnSpark,
            1e10,
            priceConverter.sDaiToEth(supplyOnSpark).mulWadDown(spark.getMaxLtv() - 0.005e18) // -0.5% to avoid borrowing at max ltv
        );

        uint256 initialBalance = supplyOnSpark.divWadDown(1e18 - floatPercentage);
        uint256 minFloat = supplyOnSpark.mulWadDown(floatPercentage);

        deal(address(sDai), address(vault), initialBalance);

        bytes[] memory callData = new bytes[](2);
        callData[0] = abi.encodeWithSelector(scDAI.supply.selector, spark.id(), supplyOnSpark);
        callData[1] = abi.encodeWithSelector(scDAI.borrow.selector, spark.id(), borrowOnSpark);

        vault.rebalance(callData);

        _assertCollateralAndDebt(spark.id(), supplyOnSpark, borrowOnSpark);
        assertApproxEqAbs(vault.totalAssets(), initialBalance, 1e10, "total assets");
        assertApproxEqRel(vault.sDaiBalance(), minFloat, 0.05e18, "float");
    }
```